airflow/api_connexion/endpoints/task_instance_endpoint.py

```diff
@@ -545,3 +546,54 @@ def post_set_task_instances_state(*, dag_id: str, session: Session = NEW_SESSION
         session=session,
     )
     return task_instance_reference_collection_schema.dump(TaskInstanceReferenceCollection(task_instances=tis))
+
+
+@security.requires_access(
+    [
+        (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG),
+        (permissions.ACTION_CAN_EDIT, permissions.RESOURCE_DAG_RUN),
+    ],
+)
+@provide_session
+def set_task_instance_note(
+    *, dag_id: str, dag_run_id: str, task_id: str, session: Session = NEW_SESSION
+) -> APIResponse:
+    """Set the note for a dag run."""
+    try:
+        post_body = set_task_instance_note_form_schema.load(get_json_request_dict())
+        new_value_for_notes = post_body["notes"]
+        # Note: We can't add map_index to the url as subpaths can't start with dashes.
+        map_index = post_body["map_index"]
+    except ValidationError as err:
+        raise BadRequest(detail=str(err))
+
+    query = (
+        session.query(TI)
+        .filter(TI.dag_id == dag_id)
+        .filter(TI.run_id == dag_run_id)
+        .filter(TI.task_id == task_id)
+    )
+    if map_index == -1:
+        query = query.filter(or_(TI.map_index == -1, TI.map_index is None))
+    else:
+        query = query.filter(TI.map_index == map_index)
+
+    try:
+        ti: TI | None = query.one_or_none()
+    except MultipleResultsFound:
+        raise NotFound(
+            "Task instance not found", detail="Task instance is mapped, add the map_index value to the URL"
+        )
+    if ti is None:
+        error_message = f'Task Instance not found for dag_id={dag_id}, run_id={dag_run_id}, task_id={task_id}'
+        raise NotFound(error_message)
+
+    ti.notes = new_value_for_notes
+    session.commit()
+
+    if map_index == -1:
+        return get_task_instance(dag_id=dag_id, dag_run_id=dag_run_id, task_id=task_id)
```

Comment:
Yes I tried this but I couldn't get this to work. The moment I add the joins I get a `Row` object back instead of an ORM `TaskInstance` object. Which will lead to the following error;
```
    ti.notes = new_value_for_notes
    session.commit()
    return task_instance_schema.dump(ti)
AttributeError: 'Row' object has no attribute 'notes'
```

Reply:
I tested the line a have put above and it worked. I had the same issue when I tried, the reason for that is we need to pass a tuple to the dump call. (The task instance schema is different from what we may expect, in a sense that it does need more that just the TI for the dump operation)

Reply:
If you want to use the same query as in `get_task_instance`:
```python
    query = (
        session.query(TI)
        .filter(TI.dag_id == dag_id, TI.run_id == dag_run_id, TI.task_id == task_id)
        .join(TI.dag_run)
        .outerjoin(
            SlaMiss,
            and_(
                SlaMiss.dag_id == TI.dag_id,
                SlaMiss.execution_date == DR.execution_date,
                SlaMiss.task_id == TI.task_id,
            ),
        )
        .add_entity(SlaMiss)
        .options(joinedload(TI.rendered_task_instance_fields))
    )
    if map_index == -1:
        query = query.filter(or_(TI.map_index == -1, TI.map_index is None))
    else:
        query = query.filter(TI.map_index == map_index)

    try:
        ti, sla_miss = query.one_or_none()
    except MultipleResultsFound:
        raise NotFound(
            "Task instance not found", detail="Task instance is mapped, add the map_index value to the URL"
        )
    if ti is None:
        error_message = f'Task Instance not found for dag_id={dag_id}, run_id={dag_run_id}, task_id={task_id}'
        raise NotFound(error_message)

    ti.notes = new_value_for_notes
    session.commit()

    return task_instance_schema.dump((ti, sla_miss))

```

Reply:
That works perfectly! Thanks. I completely missed the unpacking that happened.